onnxruntime/core/graph/graph.cc

```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
+    }
+#endif
+  }
+
+  // We want to make sure we get nodes in topological order
+  // because Constant folding may cause the nodes appear in
+  // a different order.
+  GraphViewer graph(graph_to_inline);
+  for (const auto node_idx : graph.GetNodesInTopologicalOrder()) {
+    const auto* node = graph_to_inline.GetNode(node_idx);
+    if (node == nullptr || node->OpType() == kConstant) {
+      continue;
+    }
+
+    InlinedVector<NodeArg*> new_node_input_defs;
+    for (const auto* input_def : node->InputDefs()) {
+      // Check if this is one of the implicit graph inputs
+      // then leave the name as is and reuse the NodeArg
+      const auto& input_name = input_def->Name();
+      if (if_all_inputs.count(input_name) > 0) {
+        auto* node_arg = GetNodeArgIncludingParentGraphs(input_name);
+        ORT_ENFORCE(node_arg != nullptr,
+                    "Implicit or explicit input is expected to have existing node_arg: ", input_name);
+        new_node_input_defs.push_back(node_arg);
+      } else {
+        auto hit = name_mapping.find(input_name);
+        if (hit != name_mapping.cend()) {
+          // This is other node output, constant node or initializer that was renamed.
+          auto* node_arg = GetNodeArg(hit->second);
+          ORT_ENFORCE(node_arg != nullptr,
+                      "Expecting to exist node_arg: ", input_name);
+          new_node_input_defs.push_back(node_arg);
+        } else {
+          ORT_THROW("Node's: ", node->Name(), " input: ", input_name, " is not If node's input or previous node output in this subgraph");
+        }
+      }
+    }
+
+    InlinedVector<NodeArg*> new_node_output_defs;
+    for (const auto* output_def : node->OutputDefs()) {
+      const auto& output_name = output_def->Name();
+      auto hit = name_mapping.find(output_name);
+      if (hit != name_mapping.cend()) {
+        // This is one of the graph outputs, we rename it to
+        // If node output.
+        auto* node_arg = GetNodeArg(hit->second);
+        ORT_ENFORCE(node_arg != nullptr,
+                    "Expecting to exist output node_arg: ", output_name);
+        new_node_output_defs.push_back(node_arg);
+      } else {
+        // We generate an output to downstream nodes.
+        auto new_name = GenerateNodeArgName(make_unique(output_name));
+        new_node_output_defs.push_back(&GetOrCreateNodeArg(new_name, output_def->TypeAsProto()));
+        ORT_IGNORE_RETURN_VALUE(name_mapping.emplace(output_name, std::move(new_name)));
+      }
+    }
+
+    const auto new_node_name = GenerateNodeName(make_unique(node->Name()));
+    if (node->ContainsSubgraph()) {
+      // Make a copy
+      auto attributes = node->GetAttributes();
+      RenameNodeAttributesSubgraphDependentNames(name_mapping, attributes);
+      AddNode(new_node_name, node->OpType(), node->Description(),
+              new_node_input_defs,
+              new_node_output_defs,
+              &attributes,
+              node->Domain());
```

Comment:
Feels like it adds a lot of complication to go back to updating protobuf based types when all the subgraphs would have onnxruntime::Graph instances already, and we could recurse into those to do naming updates more easily. 

Can we instead 
- recurse into the subgraphs returned by Node::GetAttributeNameToMutableSubgraphMap() (need to use that to get mutable Graph instances) and update names directly
  - we only need to recurse if an implicit input of the node containing the subgraph/s is present in the renaming map

If we do that, I don't think the node attributes need to be updated at all (a std::move would be ideal to avoid the copy cost)  The only time the GraphProto attributes matter is if we're serializing the updated model via Graph::ToGraphProto, and there's existing logic in Node::ToProto to update the subgraph attributes based on their Graph instances (`update_subgraphs` param) when that happens.